MutationObserver doesn't observe style attribute change when resizing element

annevk opened the pull request:
#### 270c2df6a704feea9fa78e1bc395f372c318fbfb
<pre>
MutationObserver doesn&apos;t observe style attribute change when resizing element
<a href="https://bugs.webkit.org/show_bug.cgi?id=266846">https://bugs.webkit.org/show_bug.cgi?id=266846</a>
<a href="https://rdar.apple.com/120109181">rdar://120109181</a>

Reviewed by Antti Koivisto.

RenderLayer::resize() did not use a StyleAttributeMutationScope and as
such style attribute changes would not be enqueued to mutation
observers.

Thanks to Antti for highlighting the likely fix.

* LayoutTests/fast/css/resize-mutation-observer-expected.txt: Added.
* LayoutTests/fast/css/resize-mutation-observer.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::resize):

Canonical link: <a href="https://commits.webkit.org/287833@main">https://commits.webkit.org/287833@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7a180220c916cd1da60f7193ded3943953a259e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80877 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/399 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85405 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31861 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82988 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/417 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8199 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63157 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20928 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83946 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/223 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73611 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43460 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27771 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30319 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71690 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28309 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86839 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8104 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5724 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71457 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8281 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69447 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70697 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17620 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14726 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13651 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8066 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13587 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7905 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11424 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on LayoutTests/fast/css/resize-mutation-observer-expected.txt:
```diff
@@ -0,0 +1,13 @@
+          text run at (0,15) width 0: " "
+          text run at (0,30) width 180: "attributes, style, null"
+layer at (8,8) size 173x125 clip at (10,10) size 169x121
+  RenderBlock {DIV} at (0,0) size 173x125 [border: (2px solid #0000FF)]
```
Please don't make new render tree dump tests. They are platform dependent and generally a pain. Instead use text dumps (`window.testRunner?.dumpAsText()`) or reftests.

anttijk commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2947,6 +2949,7 @@ void RenderLayer::resize(const PlatformMouseEvent& evt, const LayoutSize& oldOff
         styledElement->setInlineStyleProperty(CSSPropertyHeight, roundToInt(baseHeight + difference.height()), CSSUnitType::CSS_PX);
     }
 
+    mutationScope.enqueueMutationRecord();
```
You should call this in the branches that invoke `setInlineStyleProperty()` (calling it multiple times is fine) to avoid unnecessary work.

anttijk commented on LayoutTests/fast/css/resize-mutation-observer.html:
```diff
@@ -0,0 +1,40 @@
+    for (const r of records) {
+      result.textContent += `\n\n${r.type}, ${r.attributeName}, ${r.oldValue}`;
+    }
+    window.testRunner.dumpAsText(result.textContent);
```
dumpAsText() does not take arguments, you just call it in the beginning of the test and all the text content gets dumped.

Also it is good to call it as `window.testRunner?.dumpAsText()` so it doesn't blow up when running the test outside the test runner.

annevk committed:
MutationObserver doesn't observe style attribute change when resizing element
https://bugs.webkit.org/show_bug.cgi?id=266846
rdar://120109181

Reviewed by Antti Koivisto.

RenderLayer::resize() did not use a StyleAttributeMutationScope and as
such style attribute changes would not be enqueued to mutation
observers.

Thanks to Antti for highlighting the likely fix.

* LayoutTests/fast/css/resize-mutation-observer-expected.txt: Added.
* LayoutTests/fast/css/resize-mutation-observer.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::resize):

Canonical link: https://commits.webkit.org/287833@main
